### PR TITLE
test: mark difference in handling IFs in LWT as cassandra_bug 

### PR DESCRIPTION
**Pull request by andrzej-jackowski-scylladb**
Backport not needed, only docs and minor tests changes.

### CODE REVIEW

**Comment by nyh** [comment]
If we decide that this is the solution to #13011 (and nothing else needs to be done for that issue), this PR should have "Fixes #13011" so the issue will be automatically closed when it is merged.

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks for looking into this. I left some minor comments about the documentation, but more importantly I think you changed the test too much and it's not clear what is being tested now - see comments inline.

**Comment by andrzej-jackowski-scylladb** [comment]
v2:
 - Reverted old scenario of `test_lwt_with_batch_conflict_1`. Modified it to be `scylla_only` case that passes on Scylla and fails on Cassandra. Adjusted the comment.
 - Added test_lwt_with_batch_conflict_3 that is `scylla_only` case with mixed `IF expr` and `IF EXISTS`
 - Moved "Mixing LWT IF clauses in BATCH statements" section to docs/kb/lwt-differences.rst
 - Changed "ScyllaDb" to "ScyllaDB"
 - Added links to `LWT` and `lwt-differences.rst` in the relevant section of cql-extensions.md

File: test/cqlpy/test_lwt.py [hunk 1]
```diff
@@ -79,17 +79,15 @@ def test_lwt_empty_clustering_range(cql, table1):
...
-# so if a batch has both a IF EXISTS and IF NOT EXISTS on the same row, they
-# can't possibly both be true, so this batch is guaranteed to fail
-# regardless of the data. Cassandra detects this specific conflict, and
-# prints an error instead of silently failing the batch.
-# Reproduces #13011.
-@pytest.mark.xfail(reason="issue #13011")
-def test_lwt_with_batch_conflict_1(cql, table1):
+# applied. Cassandra detects when a batch contains mixed IFs and disallows
+# such request, even though it can be finished successfully.
+# Refs. #13011.
```
[the remark this answers is not in the document]
Comment (andrzej-jackowski-scylladb) [reply test/cqlpy/test_lwt.py]: Fixed

File: docs/kb/lwt-differences.rst [hunk 2]
```diff
@@ -9,8 +9,8 @@ How is it different?
 * ScyllaDB will automatically use synchronous commit log write mode for all lightweight transaction writes. Before a lightweight transaction completes, scylla will ensure that the data in it has hit the device. This is done in all commitlog_sync modes.
 * Conditional statements return a result set, and unlike Cassandra, ScyllaDB result set metadata doesn’t change from execution to execution: ScyllaDB always returns  the old version of the  row, regardless of whether the condition is true or not. This ensures conditional statements work well with prepared statements.
 * For batch statement, the returned result set contains an old row for every conditional statement in the batch, in statement order. Cassandra returns results in clustering key order.
+* For batch statement, ScyllaDB allows mixing `IF EXISTS`, `IF NOT EXISTS`, and other conditions.
```
[the remark this answers is not in the document]
Comment (andrzej-jackowski-scylladb) [reply docs/kb/lwt-differences.rst:12]: I agree, I moved the section to lwt-differences.rst. There is also existing section in cql-extensions.md regarding using NULL in LWTs. I added the links there. Maybe, there is a better option, but I don't have any good idea how to do it without a large reorganization in cql-extensions.md.

File: test/cqlpy/test_lwt.py [hunk 3]
```diff
@@ -79,17 +79,15 @@ def test_lwt_empty_clustering_range(cql, table1):
...
+def test_lwt_with_batch_conflict_1(cql, table1, cassandra_bug):
     p = unique_key_int()
-    with pytest.raises(InvalidRequest, match='Cannot mix'):
-        cql.execute(f'BEGIN BATCH DELETE FROM {table1} WHERE p={p} AND c=1 IF EXISTS; INSERT INTO {table1}(p,c,r) VALUES ({p},1,2) IF NOT EXISTS; APPLY BATCH;')
+    cql.execute(f'INSERT INTO {table1}(p,c,r) VALUES ({p},1,1)')
```
[the remark this answers is not in the document]
Comment (andrzej-jackowski-scylladb) [reply test/cqlpy/test_lwt.py]: I checked it again, and Cassandra **does allow** this case (I made a mistake earlier, because I missed the fact that run-cassandra marks successful tests as XFAIL instead of XPASS). Therefore, I reverted to the original test scenario that changes the same row.

**Comment by andrzej-jackowski-scylladb** [comment]
@nyh, all your remarks have been addressed

File: docs/kb/lwt-differences.rst [hunk 4]
```diff
@@ -9,8 +9,55 @@ How is it different?
...
+
+Additional Notes
+================
+
+Mixing LWT IF clauses in BATCH statements
+-----------------------------------------
+
+`Conditional batches` are `BATCH` statements that contain one or more conditional statements.
+A batch is executed only if all conditions in all statements are true, and all conditions are evaluated against the initial database state.
+ScyllaDB allows using different conditions such as `IF EXISTS`, `IF NOT EXISTS`, and `IF expression`, in statements comprising a batch. Some examples:
```
Comment (nyh) [inline docs/kb/lwt-differences.rst:27]: Unfortunately, this explanation is misleading, I'm guessing you wrote it before our previous discussion crystalized your understanding of what is really happening here.

It is **always** allowed, in both Cassandra and Scylla to have multiple statements in the same batch with different types of conditions if they go to different clustering rows (in the same partition). The only question is what happens if you have two of the statement in the batch going to the **same** row. Clearly, if the IF conditions are different for the same row, one of them _must_ fail and the entire batch will fail (not be done). The Cassandra developers decided that in some of these cases, this conflict is probably a typo (e.g., why would anyone deliberately have an IF EXISTS and IF NOT EXISTS for the same row?) and it will be helpful to make this into an invalid request - not just a failed batch. But Cassandra did this inconsistently (catch some conflicts, not others) and in Scylla we decided not to attempt such detection at all - whenever a batch contains two conflicting conditions on the same row, one evaluates to true and the other to false, the batch will fail normally - it will not be an invalid request.

File: test/cqlpy/test_lwt.py [hunk 5]
```diff
@@ -83,13 +83,18 @@ def test_lwt_empty_clustering_range(cql, table1):
...
+# we considered this check to be inconsistent and unhelpful, and
+# decided not to implement it, and this case is treated as a normal batch
+# failure (not all conditions are true), not an error. See discussion in #13011.
+# The test is marked scylla_only because it will fail on Cassandra which will
+# report an error, not a batch failure.
+def test_lwt_with_batch_conflict_1(cql, table1, scylla_only):
     p = unique_key_int()
-    with pytest.raises(InvalidRequest, match='Cannot mix'):
-        cql.execute(f'BEGIN BATCH DELETE FROM {table1} WHERE p={p} AND c=1 IF EXISTS; INSERT INTO {table1}(p,c,r) VALUES ({p},1,2) IF NOT EXISTS; APPLY BATCH;')
+    rs = list(cql.execute(f'BEGIN BATCH DELETE FROM {table1} WHERE p={p} AND c=1 IF EXISTS; INSERT INTO {table1}(p,c,r) VALUES ({p},1,2) IF NOT EXISTS; APPLY BATCH;'))
```
Comment (nyh) [inline test/cqlpy/test_lwt.py:94]: I think the `list()` isn't necessary here (the iterator can be iterated without being materialized first). But I guess it doesn't hurt either. But I see you copied this line from a pre-existing function, so fine.

File: test/cqlpy/test_lwt.py [hunk 6]
```diff
@@ -103,6 +108,17 @@ def test_lwt_with_batch_conflict_2(cql, table1):
...
+# Moreover, there are cases where Cassandra prevents mixing
+# IF EXISTS with different conditions such as IF r=1, despite
+# the fact that the conditions are not contradictory.
+def test_lwt_with_batch_conflict_3(cql, table1, scylla_only):
+    p = unique_key_int()
+    cql.execute(f'INSERT INTO {table1}(p,c,r) VALUES ({p},1,1)')
+    rs = list(cql.execute(f'BEGIN BATCH UPDATE {table1} SET r=10 WHERE p={p} AND c=1 IF r=1; UPDATE {table1} SET r=20 WHERE p={p} AND c=1 IF EXISTS;  APPLY BATCH;'))
+    # Cassandra fails with: "Cannot mix IF conditions and IF EXISTS for the same row"
+    for r in rs:
+        assert r.applied == True
```
Comment (nyh) [inline test/cqlpy/test_lwt.py:120]: Nice catch. It's weird that Cassandra won't apply this **successful** batch. I wonder if it's not even a bug that should be reported to them. But never mind :-)

**Review by nyh — CHANGES_REQUESTED** [review]
Looks very good, but I think that your documentation patch still reflects your old, inaccurate, understanding of the problem and should be updated.

**Comment by andrzej-jackowski-scylladb** [comment]
v3:
- Modified the documentation to emphasize that conditions cannot be mixed for the same row
- Removed unneeded list() from new test cases
- Added test_lwt_with_batch_conflict_4 and test_lwt_with_batch_conflict_5 to present how mixing conditions works with static rows

File: docs/kb/lwt-differences.rst (hunk 4 above)
Comment (andrzej-jackowski-scylladb) [reply docs/kb/lwt-differences.rst]: Fixed

**Comment by andrzej-jackowski-scylladb** [comment]
@nyh, I also added two new test cases (4 and 5) because for me the behavior of static columns was confusing. I'm not sure if static columns should be mentioned in `lwt-differences.rst`, most likely it is too specific.

**Comment by andrzej-jackowski-scylladb** [comment]
@nyh, sorry for the large number of pushes with cosmetic changes in test comments. I should have slowed down instead of pushing quickfixes.

File: test/cqlpy/test_lwt.py [hunk 7]
```diff
@@ -103,6 +108,40 @@ def test_lwt_with_batch_conflict_2(cql, table1):
...
+
+# During a static column update, the clustering key is not required in the WHERE clause.
+# A batch composed of a query that updates a static column and a query that inserts
+# a new row is allowed and will be applied successfully.
+def test_lwt_with_batch_conflict_4(cql, table1):
+    p = unique_key_int()
+    cql.execute(f'INSERT INTO {table1}(p,c,r,s) VALUES ({p},1,1,1)')
+    rs = cql.execute(f'BEGIN BATCH UPDATE {table1} SET s=NULL WHERE p={p} IF EXISTS; INSERT INTO {table1}(p,c,r) VALUES ({p},2,2) IF NOT EXISTS; APPLY BATCH;')
+    for r in rs:
+        assert r.applied == True
```
Comment (nyh) [inline test/cqlpy/test_lwt.py:130]: I don't understand why this test shows any sort of "conflict". I'm not against this test though I'm not sure I understand why it's useful - but perhaps you should reconsider its name?

File: test/cqlpy/test_lwt.py [hunk 8]
```diff
@@ -103,6 +108,40 @@ def test_lwt_with_batch_conflict_2(cql, table1):
...
+# is disallowed in Cassandra.
+def test_lwt_with_batch_conflict_5(cql, table1, scylla_only):
+    p = unique_key_int()
+    cql.execute(f'INSERT INTO {table1}(p,c,r,s) VALUES ({p},1,1,1)')
+    rs = cql.execute(f'BEGIN BATCH UPDATE {table1} SET s=NULL WHERE p={p} IF EXISTS; UPDATE {table1} SET s=NULL WHERE p={p} IF s = 1; APPLY BATCH;')
+    # Cassandra fails with: "Cannot mix IF conditions and IF NOT EXISTS for the same row"
+    # The fact that "IF NOT EXISTS" is used in the error message is a bit misleading,
+    # but that's what Cassandra returns
+    for r in rs:
+        assert r.applied == True
```
Comment (nyh) [inline test/cqlpy/test_lwt.py:142]: Interesting, in this, like test_lwt_with_batch_conflict_3, it's not clear to me if this is a missing Cassandra feature or an outright bug - if it's the latter we should prefer the "cassadra_bug" label over the "scylla_only" one (the difference is mostly cosmetic).

If you have the inclination to do so, you can even report these suspected bugs to the Cassandra project, and see what they think. Maybe they'll fix it, or maybe they can explain (?) how their weird behavior is actually correct or necessary for some reason.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
> @nyh, I also added two new test cases (4 and 5) because for me the behavior of static columns was confusing. I'm not sure if static columns should be mentioned in `lwt-differences.rst`, most likely it is too specific.

I feel like there are two separate types of "differences" here:

* One thing is Cassandra's attempt to "help you" by detecting tautological conflicts in conditions (a set of conditions that can never succeed together regardless of the data) and reporting these as errors - which we don't do. I think this is a real difference that we should explain (and you did).
* A second thing is Cassandra's outright bugs (I believe) in doing the above correctly, leading it to report "conflicts" in at least two cases you found (one of them involving static columns) which clearly have no conflict. I think when we consider those two things bugs in Cassandra, we don't need to mention them in lwt-differences.rst - we don't need to list their outright bugs that we fixed. However, if you have the energy you might want to report this bug (with its two examples)  to the Cassandra developers - maybe they will acknowledge it is a bug and fix it (so we'll not have this difference) or maybe they will convince you (?) that their weird behavior is actually the right (?) thing. But perhaps we spent too much time on this issue already and you shouldn't bother yourself to report it to Cassandra. Your choice.

**Comment by nyh** [comment]
I'll merge this PR as-is. We can probably continue to debate some finer point of the test naming and why Cassandra behaves in certain ways for even longer, but I think we already spent quite a bit of time on this not-super-important corner case and the PR is already in a good enough shape. If you later want to continue renaming tests or adding more tests or whatever, feel free to send a followup patch.
